### DISCUSSION TITLE
Exposing first hit CastRay

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#50] [Ideas]
It would be great to have a CastRay for the first hit exposed as well, as it is used as often as its slower brother. Perhaps under a different name, e.g. CastRayFirst?

**Comment by jrouwe** [comment]
To get only the first hit you would update the early out fraction as shown here:

https://github.com/jrouwe/JoltPhysics.js/blob/946114bfd0f82bbb63a25f2c1f61feb2b1aa18b5/Examples/ray_cast.html#L63

The hits are processed roughly from near to far, and as soon as you set the early out fraction to a particular value, you will not receive any hits that are further away than that value. Since the sorting is not perfect, you do need to update your collision result if you receive another callback to AddHit with a closer fraction.

**Comment by jrouwe** [comment]
I also exposed a convenience class for finding the first hit ```CastRayClosestHitCollisionCollector``` and update the example:

https://github.com/jrouwe/JoltPhysics.js/blob/9c086477c412fe42f7e30c7e938bd9ddd9a0bfce/Examples/ray_cast.html#L90-L97

**Comment by LeXXik** [comment]
That's great, thank you!
